rp2040-hal/src/multicore.rs

```diff
@@ -157,14 +180,19 @@ impl<'p> Core<'p> {
             // Set up the stack
             let mut stack_ptr = unsafe { stack.as_mut_ptr().add(stack.len()) };
 
-            let mut push = |v: usize| unsafe {
+            // We don't want to drop this, since it's getting moved to the other core.
+            let mut entry = ManuallyDrop::new(entry);
+
+            // Push the arguments to `core1_startup` onto the stack.
+            unsafe {
+                // Push `stack_bottom`.
                 stack_ptr = stack_ptr.sub(1);
-                stack_ptr.write(v);
-            };
+                stack_ptr.cast::<*mut usize>().write(stack.as_mut_ptr());
 
-            push(wrapper as usize);
-            push(stack.as_mut_ptr() as usize);
-            push(entry as usize);
+                // Push `entry`.
+                stack_ptr = stack_ptr.sub(1);
+                stack_ptr.cast::<&mut ManuallyDrop<F>>().write(&mut entry);
+            }
```

Comment:
To be pedantic: Do we need a memory barrier at this point, to make sure the new stack values are actually written before core1 tries to load them?

Reply:
If we're being pedantic, yes.
A compiler fence is probaby enough but we might as well do it properly.

Reply:
```suggestion
            }
            // Make sure the compiler does not reorder the stack writes
            // in relation to the fifo writes, below.
            //
            // From the compiler perspective, this doesn't guarantee that the
            // second core actually sees those writes. However, we know that
            // the RP2040 doesn't have memory caches, and writes happen in-order.
            compiler_fence(Ordering::AcqRel);

```
(And add `use core::sync::atomic::{compiler_fence, Ordering};` above).

I did not use `core::sync::atomic::fence` because that doesn't provide additional guarantees without an additional atomic variable shared between the cores, as explained in https://doc.rust-lang.org/core/sync/atomic/fn.fence.html.

Reply:
Why does this need `Ordering::AcqRel`? To me, it seems like `Release` should be enough.

Reply:
Good question. You are right, `Release` should be enough.
My reasoning was that I also want to prevent later accesses from being reordered to before this fence. But that's actually already guaranteed by `Release`: "with Release, preceding reads and writes cannot be moved past subsequent writes"  implies that subsequent writes can't be moved to before preceding writes, otherwise it doesn't make any sense.

```suggestion
            }
            // Make sure the compiler does not reorder the stack writes
            // in relation to the fifo writes, below.
            //
            // From the compiler perspective, this doesn't guarantee that the
            // second core actually sees those writes. However, we know that
            // the RP2040 doesn't have memory caches, and writes happen in-order.
            compiler_fence(Ordering::Release);

```